notation/notation.go

```diff
@@ -0,0 +1,285 @@
+*/
+func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (ocispec.Descriptor, []*VerificationOutcome, error) {
+	artifactRef := opts.ArtifactReference
+	trustpolicyDoc, err := verifier.TrustPolicyDocument()
+	if err != nil {
+		return ocispec.Descriptor{}, nil, ErrorNoApplicableTrustPolicy{Msg: err.Error()}
+	}
+	trustPolicy, err := trustpolicyDoc.GetApplicableTrustPolicy(artifactRef)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, ErrorNoApplicableTrustPolicy{Msg: err.Error()}
+	}
+
+	var verificationOutcomes []*VerificationOutcome
+	// ignore the error since we already validated the policy document
+	verificationLevel, _ := trustPolicy.SignatureVerification.GetVerificationLevel()
+	if verificationLevel.Name == trustpolicy.LevelSkip.Name {
+		verificationOutcomes = append(verificationOutcomes, &VerificationOutcome{VerificationLevel: verificationLevel})
+		return ocispec.Descriptor{}, verificationOutcomes, nil
+	}
+
+	// get signature manifests
+	var success bool
+	var verifiedSigBlobDesc ocispec.Descriptor
+	artifactDescriptor, err := repo.Resolve(ctx, artifactRef)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, ErrorSignatureRetrievalFailed{Msg: err.Error()}
+	}
+	err = repo.ListSignatures(ctx, artifactDescriptor, func(signatureManifests []ocispec.Descriptor) error {
+		// if already verified successfully, no need to continue
+		if success {
+			return nil
+		}
+
+		if len(signatureManifests) < 1 {
+			return ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("no signatures are associated with %q, make sure the image was signed successfully", artifactRef)}
+		}
+
+		// process signatures
+		for _, sigManifest := range signatureManifests {
+			// get signature envelope
+			sigBlob, sigBlobDesc, err := repo.FetchSignatureBlob(ctx, sigManifest)
+			if err != nil {
+				return ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("unable to retrieve digital signature with digest %q associated with %q from the registry, error : %s", sigBlobDesc.Digest, artifactRef, err.Error())}
+			}
+			_, outcome, err := verifier.Verify(ctx, sigBlob, opts)
+			if err != nil {
+				if outcome != nil {
+					outcome.SignatureBlobDescriptor = &sigBlobDesc
+					outcome.Error = err
+					verificationOutcomes = append(verificationOutcomes, outcome)
+				}
+				continue
+			}
+
+			// At this point, we've found a signature verified successfully
+			outcome.SignatureBlobDescriptor = &sigBlobDesc
+			verificationOutcomes = append(verificationOutcomes, outcome)
+			payload := &payload{}
+			err = json.Unmarshal(outcome.EnvelopeContent.Payload.Content, payload)
+			if err != nil {
+				outcome.Error = err
+				continue
+			}
+			outcome.SignedAnnotations = payload.TargetArtifact.Annotations
+			success = true
+			// Descriptor of the signature blob that gets verified successfully
+			verifiedSigBlobDesc = sigBlobDesc
+
+			return nil
```

Comment:
> Can line 171-212 moved to verifier pkg? Basically have a `verifier.Verify` method similar to `signer.Sign`. For local verification we can have method ``verifier.LocalVerify`,

That's going to be a relatively big change of the verifier interface design. @shizhMSFT any suggestion/comment/concern on this one?

Reply:
> Can line 171-212 moved to verifier pkg? Basically have a `verifier.Verify` method similar to `signer.Sign`. For local verification we can have method ``verifier.LocalVerify`,

Had a discussion with @shizhMSFT, we suggest to keep the current design. Verifying remote is based on verifying local. That's because when verifying remotely, we still need to Resolve and Fetch the remote signatures to local and then do the verification.